[lit] Fix substitutions containing backslashes

mstorsjo opened the pull request:
Substitutions can be added in a couple different ways; they can be added via the calling python scripts by adding entries to the config.substitutions dictionary, or via DEFINE lines in the scripts themselves.

The substitution strings passed to Python's re classes are interpreted so that backslashes expand to escape sequences, and literal backslashes need to be escaped.

On Unix, the script defined substitutions don't (usually, so far) contain backslashes - but on Windows, they often do, due to paths containing backslashes. This lead to a Windows specific escaping of backslashes before doing Python re substitutions - since 7c9eab8fef0ed79a5911d21eb97b6b0fa9d39f82. There's nothing inherently Windows specific about this though - any intended literal backslashes in the substitution strings need to be escaped; this is how the Python re API works.

The DEFINE lines were added later, and in order to cope with backslashes, escaping of backslashes was added in the SubstDirective class in TestRunner, applying to DEFINE lines in the tests only.

The fact that the escaping right before passing to the Python re API was done conditionally on Windows led to two inconsistencies:

- DEFINE lines in the tests that contain backslashes got double backslashes on Windows. (This was visible as a FIXME in llvm/utils/lit/tests/Inputs/shtest-define/value-escaped.txt.)

- Script provided substitutions containing backslashes did not work on Unix, but they did work on Windows.

By removing the escaping from SubstDirective and escaping it unconditionally in the processLine function, before feeding the substitutions to Python's re classes, we should have consistent behaviour across platforms, and get rid of the FIXME in the lit test.

This fixes issues with substitutions containing backslashes on Unix platforms, as encountered in PR #86649.

mstorsjo committed:
[lit] Fix substitutions containing backslashes

Substitutions can be added in a couple different ways; they can
be added via the calling python scripts by adding entries to the
config.substitutions dictionary, or via DEFINE lines in the
scripts themselves.

The substitution strings passed to Python's re classes are
interpreted so that backslashes expand to escape sequences, and
literal backslashes need to be escaped.

On Unix, the script defined substitutions don't (usually, so
far) contain backslashes - but on Windows, they often do, due to
paths containing backslashes. This lead to a Windows specific
escaping of backslashes before doing Python re substitutions -
since 7c9eab8fef0ed79a5911d21eb97b6b0fa9d39f82. There's nothing
inherently Windows specific about this though - any intended
literal backslashes in the substitution strings need to be escaped;
this is how the Python re API works.

The DEFINE lines were added later, and in order to cope with
backslashes, escaping of backslashes was added in the
SubstDirective class in TestRunner, applying to DEFINE lines in
the tests only.

The fact that the escaping right before passing to the Python re
API was done conditionally on Windows led to two inconsistencies:

- DEFINE lines in the tests that contain backslashes got double
  backslashes on Windows. (This was visible as a FIXME in
  llvm/utils/lit/tests/Inputs/shtest-define/value-escaped.txt.)

- Script provided substitutions containing backslashes did not
  work on Unix, but they did work on Windows.

By removing the escaping from SubstDirective and escaping it
unconditionally in the processLine function, before feeding the
substitutions to Python's re classes, we should have
consistent behaviour across platforms, and get rid of the FIXME
in the lit test.

This fixes issues with substitutions containing backslashes on
Unix platforms, as encountered in PR #86649.

mstorsjo commented:
Ping @jdenny-ornl

jdenny-ornl commented:
> Script provided substitutions containing backslashes did not work on Unix, but they did work on Windows.

That depends on the definition of "work".  In downstream test suites (in lit config scripts but obviously not (RE)DEFINE), I have made use of replacement string backslashes (e.g., `\g`) as defined by `re.sub`.  I doubt that specific work will ever go upstream, but it makes me wonder if we should advertise this change in behavior more broadly.

> This lead to a Windows specific escaping of backslashes before doing Python re substitutions - since [7c9eab8](https://github.com/llvm/llvm-project/commit/7c9eab8fef0ed79a5911d21eb97b6b0fa9d39f82).

I think that means my usage above cannot behave correctly in Windows, where I never tested it.  With that in mind, my estimation is that backslash escaping is the desired and safest behavior in the vast majority of cases (probably all upstream cases).  
We could surely find some per-substitution switch to opt-in to full `re.sub` behavior.

Please update
[
https://llvm.org/docs/TestingGuide.html#substitutions

  I also tried to convey this usage, perhaps not as clearly .  See the first bullet under "Test-specific substitutions" in <https://llvm.org/docs/TestingGuide.html#substitutions>:

jdenny-ornl commented:
Sorry, misclick.

jdenny-ornl commented:
Please update the first bullet under "Test-specific substitutions" in <https://llvm.org/docs/TestingGuide.html#substitutions> as it advertises `re.sub` behavior:

> Lit configuration files (e.g., lit.cfg or lit.local.cfg) can define substitutions for all tests in a test directory. They do so by extending the substitution list, config.substitutions. Each item in the list is a tuple consisting of a pattern and its replacement, which lit applies using python’s re.sub function.

The following text comes later and specifies the different behavior for (RE)DEFINE:

> Substitution value: The value includes all text from the first non-whitespace character after = to the last non-whitespace character. If there is no non-whitespace character after =, the value is the empty string. Escape sequences that can appear in python re.sub replacement strings are treated as plain text in the value.

It might also be good to have a test case showing that lit config script substitutions properly handle escape sequences.  You're already updating the one for (RE)DEFINE.

mstorsjo committed:
Update docs, test a substitution with backslashes

mstorsjo commented:
> > Script provided substitutions containing backslashes did not work on Unix, but they did work on Windows.
> 
> That depends on the definition of "work". In downstream test suites (in lit config scripts but obviously not (RE)DEFINE), I have made use of replacement string backslashes (e.g., `\g`) as defined by `re.sub`. I doubt that specific work will ever go upstream, but it makes me wonder if we should advertise this change in behavior more broadly.

Oh, that's interesting to know - especially as this case actually seems to have been documented to do this.

> > This lead to a Windows specific escaping of backslashes before doing Python re substitutions - since [7c9eab8](https://github.com/llvm/llvm-project/commit/7c9eab8fef0ed79a5911d21eb97b6b0fa9d39f82).
> 
> I think that means my usage above cannot behave correctly in Windows, where I never tested it. With that in mind, my estimation is that backslash escaping is the desired and safest behavior in the vast majority of cases (probably all upstream cases). In the future if needed, we could surely find some per-substitution switch to opt-in to full `re.sub` behavior.

Thanks! Indeed, this is probably the behaviour that is most convenient for most uses, but as you say, it may be good to be able to use the real escape sequences as well somehow.

> Please update the first bullet under "Test-specific substitutions" in https://llvm.org/docs/TestingGuide.html#substitutions as it advertises `re.sub` behavior.

Ok, I updated the PR with that docs change. This does indeed highlight that this is, in principle a breaking change wrt the docs, even if no cases upstream did exercise this (and it wasn't working as intended on Windows).

> It might also be good to have a test case showing that lit config script substitutions properly handle escape sequences. You're already updating the one for (RE)DEFINE.

Done - I added a case for that. I wasn't sure which test suite would be best to have (there didn't seem to be any subdirectory specifically for testing general substitutions), so I stuck it in `shtest-define` next to the other test - feel free to suggest other better places for it.

jdenny-ornl commented on llvm/utils/lit/tests/Inputs/shtest-define/lit.cfg:
```diff
@@ -22,6 +22,7 @@ config.substitutions.insert(0, ("%{global:greeting}", ""))
 config.substitutions.insert(
     0, ("%{global:echo}", "echo GLOBAL: %{global:greeting} %{global:what}")
 )
+config.substitutions.insert(0, ("%{global:subst-with-escapes}", r"value-with-\g"))
```
The existing comments above don't describe this substitution.

```suggestion
config.substitutions.insert(
    0, ("%{global:echo}", "echo GLOBAL: %{global:greeting} %{global:what}")
)

# This substitution includes an re.sub replacement string escape sequence, 
# which lit should treat as plain text.
config.substitutions.insert(0, ("%{global:subst-with-escapes}", r"value-with-\g"))
```

jdenny-ornl commented:
LGTM besides the one nit.  Thanks for working on this.

mstorsjo committed:
Update llvm/utils/lit/tests/Inputs/shtest-define/lit.cfg

Co-authored-by: Joel E. Denny <jdenny.ornl@gmail.com>

mstorsjo committed:
Remove trailing whitespace
